wgsl/index.bs

```diff
@@ -520,6 +520,76 @@ type.
+
+A <dfn noexport>texture</dfn> is a grid of texels supporting special operations useful for rendering.
+A texture's representation is typically optimized for those operations, with many details hidden from the programmer
+including data layouts, data types, and operations that cannot be expressed directly in the shader language.
+Each texel has a unique grid coordinate within the texture.
+
+An <dfn noexport>arrayed texture</dfn> is a texture which is organized as a homegeneous array of grids of texels.
+In an arrayed texture, a texel is associated with a unique combination of array index and grid coordinate.
+
+A texture has the following features:
+
+: <dfn noexport>texel format</dfn>
+:: the data in each texel. See [[#texel-formats]]
+:  dimensionality
+:: the number of dimensions in the grid coordinates, and how the coordinates are interpreted.
+    The number of dimensions is 1, 2, or 3.
+    In some cases the third coordinate is decomposed so as to specify a cube face and a layer index.
+: size
+:: the extent of grid coordinates along each dimension
+: arrayed
+:: whether the texture is arrayed
+: <dfn noexport>array size</dfn>
+:: the number of homogeneous grids, if the texture is arrayed
+
+A shader does not have direct access to the texel storage within a texture variable.
+Instead, texture builtin functions are used to access the data.
+
+[SHORTNAME] defines texture types which control which operations are available.
+determine the set of usable builtin functions.
+
+TODO(dneto): Describe how texture types determine whether texels are readable, writeable,
+how they are addressed, and how samplers affect both addressing and filtering.
+Preview that details are provided in later sections.
+
+### Texel formats ### {#texel-formats}
+
+A [=texel format=] specifies the data contained in a [=texel=].
+
+A texel format is characterized by:
+
+: <dfn noexport>channels</dfn>
+:: Each channel contains a scalar.
+    There are up to four channels, named `r`, `g`, `b`, or `a`.
+: <dfn noexport>channel format</dfn>
+:: The number of bits in the channel, and how those bits are interpreted.
+    In some cases different channels in a texel will have different channel types.
+
+<table class='data'>
+  <caption>Channel formats</caption>
+  <thead>
+    <tr><td>Channel format<td>Number of stored bits<td>Interpetation of stored bits<td>Shader type<td>Shader value
+  </thead>
+  <tr><td>8unorm<td>8<td>unsigned integer |v| &isinv; {0,...,255}<td>f32<td> |v| &div; 255
+  <tr><td>8snorm<td>8<td>signed integer |v| &isinv; {-128,...,127}<td>f32<td> |v| &div; 128
+  <tr><td>8uint<td>8<td>unsigned integer |v| &isinv; {0,...,255}<td>u32<td> |v| &div; 255
+  <tr><td>8sint<td>8<td>signed integer |v| &isinv; {-128,...,127}<td>i32<td> |v| &div; 128
+  <tr><td>16uint<td>16<td>unsigned integer |v| &isinv; {0,...,65535}<td>u32<td> |v| &div; 65535
+  <tr><td>16sint<td>16<td>signed integer |v| &isinv; {-32768,...,32767}<td>i32<td> |v| &div; 32768
```

Comment:
I don't think that the WGSL spec should talk about the exact bit representation of texture, because there's weird formats like rgb9e5 and compressed ones. Only the component type matters, except for storage textures (which can have an exact bit representation described in this spec).

Reply:
I think there is still value in describing the formats we *do* know and guarantee, such as `8unorm`. For `rgb9e5` we can figure something out and say it's more platform-dependent.

Reply:
I'm not sure I understand, should the WGSL spec just tell you that when you sample a texture with floating point component type you get an `f32` and what `f32` you get is described by the texture sampling part of the WebGPU specification? This way WGSL wouldn't care if it is `r8unorm` `rg16float` or `bc1-rgba-unorm`?

Reply:
well, I'm saying that it's generally useful that, for example, an r8unorm is one of the values between 0/255 and 255/255. This allows authors to adjust their expectations on sampling, or when writing to the color outputs.
Where exactly this is specified (WGSL or WebGPU spec) is not something that I feel strongly about.

Reply:
The only reason WGSL needs texel-formats is to parameterize storage textures. The sampled and depth textures do not, and I'd like to eliminate those weirdly-shaped texel formats from the WGSL spec entirely.

See the modifications to the texture-ro and texture-wo sections in this MR.

Reply:
After removing the texel formats that require SPIR-V capability StorageImageExtendedFormats, we're left with:

* r32uint
* r32sint
* r32float
* rgba8unorm
* rgba8snorm
* rgba8uint
* rgba8sint

I can delete all the other ones from the tables, and make this much simpler.

Reply:
where did 32-bit RG and RGBA formats go?

Reply:
Looks like `rgba32float` and `rgba32ui`, `rgba32i` are missing from list above but all of the `rg` formats appear to be in `StorageImageExtendedFormats`.  https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#Image_Format

Reply:
Interesting! So this is different from what Vulkan requires. I guess the `StorageImageExtendedFormats` on the SPIR-V side becomes an implementation detail, and we don't need to care about it on the API side.
Please just follow the table we have in the spec - https://gpuweb.github.io/gpuweb/#plain-color-formats

Reply:
I had been reading the required format support for Vulkan. But I have not been following what was worked out already for WebGPU. 
I'll follow the WebGPU table. Thanks!